### Dummy Variables

You saw in the earlier notebook that you weren't able to directly add a categorical variable to your multiple linear regression model. In this notebook, you will get some practice incorporating categorical data by converting to dummy variables in your models and interpreting the output.

Let's start by reading in the necessary libraries and data.

In [1]:
import numpy as np
import pandas as pd
import statsmodels.api as sm;
import matplotlib.pyplot as plt
%matplotlib inline

df = pd.read_csv('./house_prices.csv')
df.head()

/opt/conda/lib/python3.6/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


house_id neighborhood  area  bedrooms  bathrooms      style    price
0      1112            B  1188         3          2      ranch   598291
1       491            B  3512         5          3  victorian  1744259
2      5952            B  1134         3          2      ranch   571669
3      3525            A  1940         4          2      ranch   493675
4      5108            B  2208         6          4  victorian  1101539

`1.` Use the [pd.get_dummies](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html) documentation to assist you with obtaining dummy variables for the **neighborhood** column.  Then use [join](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.join.html) to add the dummy variables to your dataframe, **df**, and store the joined results in **df_new**.

Fit a linear model using **all three levels** of **neighborhood** to predict the price. Don't forget an intercept.

Use your results to answer quiz 1 below.

In [2]:
df1  = pd.get_dummies(df['neighborhood'])


In [3]:
df_new = df.join(df1, how = 'right')
df_new.head()

house_id neighborhood  area  bedrooms  bathrooms      style    price  A  B  \
0      1112            B  1188         3          2      ranch   598291  0  1   
1       491            B  3512         5          3  victorian  1744259  0  1   
2      5952            B  1134         3          2      ranch   571669  0  1   
3      3525            A  1940         4          2      ranch   493675  1  0   
4      5108            B  2208         6          4  victorian  1101539  0  1   

   C  
0  0  
1  0  
2  0  
3  0  
4  0

`2.`  Now, fit an appropriate linear model for using **neighborhood** to predict the price of a home. Use **neighborhood A** as your baseline. (And remember that the values shown in the results for the other neighborhoods will be based on comparisons with this baseline neighborhood A then.) Use your resulting model to answer the questions in Quiz 2 and Quiz 3 below.

In [4]:
df_new['intercept'] = 1
lm = sm.OLS(df_new['price'], df_new[['intercept', 'B', 'C']])
results = lm.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.246
Model:                            OLS   Adj. R-squared:                  0.246
Method:                 Least Squares   F-statistic:                     983.1
Date:                Sun, 31 Jan 2021   Prob (F-statistic):               0.00
Time:                        16:09:19   Log-Likelihood:                -87082.
No. Observations:                6028   AIC:                         1.742e+05
Df Residuals:                    6025   BIC:                         1.742e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
intercept   5.411e+05   1.05e+04     51.537      0.000    5.21e+05    5.62e+05
B           5.295e+05    1.4e+04     37.870      0.000    5.02e+05    5.57e+05
C           -332.3594   1.52e+04     -0.022      0.983   -3.01e+04    2.94e+04
==============================================================================
Omnibus:                      689.315   Durbin-Watson:                   1.999
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1154.155
Skew:                           0.793   Prob(JB):                    2.39e-251
Kurtosis:                       4.442   Cond. No.                         3.88
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

`3.` Run the two cells below to look at the home prices for the A and C neighborhoods. Add neighborhood B. This creates a glimpse into the differences that you found in the previous linear model.

In [5]:
lm = sm.OLS(df_new['price'], df_new[['intercept', 'A', 'C']])
results = lm.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.246
Model:                            OLS   Adj. R-squared:                  0.246
Method:                 Least Squares   F-statistic:                     983.1
Date:                Sun, 31 Jan 2021   Prob (F-statistic):               0.00
Time:                        16:09:25   Log-Likelihood:                -87082.
No. Observations:                6028   AIC:                         1.742e+05
Df Residuals:                    6025   BIC:                         1.742e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
intercept   1.071e+06   9231.431    115.972      0.000    1.05e+06    1.09e+06
A          -5.295e+05    1.4e+04    -37.870      0.000   -5.57e+05   -5.02e+05
C          -5.298e+05   1.43e+04    -36.991      0.000   -5.58e+05   -5.02e+05
==============================================================================
Omnibus:                      689.315   Durbin-Watson:                   1.999
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1154.155
Skew:                           0.793   Prob(JB):                    2.39e-251
Kurtosis:                       4.442   Cond. No.                         3.46
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [6]:
lm = sm.OLS(df_new['price'], df_new[['intercept', 'A', 'B', 'C']])
results = lm.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.246
Model:                            OLS   Adj. R-squared:                  0.246
Method:                 Least Squares   F-statistic:                     655.0
Date:                Sun, 31 Jan 2021   Prob (F-statistic):               0.00
Time:                        16:09:30   Log-Likelihood:                -87082.
No. Observations:                6028   AIC:                         1.742e+05
Df Residuals:                    6024   BIC:                         1.742e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
intercept   -7.01e+17   8.86e+17     -0.792      0.429   -2.44e+18    1.04e+18
A            7.01e+17   8.86e+17      0.792      0.429   -1.04e+18    2.44e+18
B            7.01e+17   8.86e+17      0.792      0.429   -1.04e+18    2.44e+18
C            7.01e+17   8.86e+17      0.792      0.429   -1.04e+18    2.44e+18
==============================================================================
Omnibus:                      682.618   Durbin-Watson:                   1.999
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1140.081
Skew:                           0.788   Prob(JB):                    2.72e-248
Kurtosis:                       4.435   Cond. No.                     3.50e+14
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 6.59e-26. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

`4.` Now, add dummy variables for the **style** of house. Create a new linear model using these new dummies, as well as the previous **neighborhood** dummies.  Use **ranch** as the baseline for the **style**.  Additionally, add **bathrooms** and **bedrooms** to your linear model.  Don't forget an intercept.  Use the results of your linear model to answer the last two questions below. **Home prices are measured in dollars, and this dataset is not real.**

To minimize scrolling, it might be useful to open another browser window to this concept to answer the quiz questions.

In [7]:
df2 = pd.get_dummies(df_new['style'])
df2_new = df_new.join(df2, how = 'right')
df2_new.head()


house_id neighborhood  area  bedrooms  bathrooms      style    price  A  B  \
0      1112            B  1188         3          2      ranch   598291  0  1   
1       491            B  3512         5          3  victorian  1744259  0  1   
2      5952            B  1134         3          2      ranch   571669  0  1   
3      3525            A  1940         4          2      ranch   493675  1  0   
4      5108            B  2208         6          4  victorian  1101539  0  1   

   C  intercept  lodge  ranch  victorian  
0  0          1      0      1          0  
1  0          1      0      0          1  
2  0          1      0      1          0  
3  0          1      0      1          0  
4  0          1      0      0          1

In [36]:
#df2_new.drop(columns = 'ranch', axis = 1)

In [8]:
lm = sm.OLS(df2_new['price'], df2_new[['intercept', 'bedrooms', 'bathrooms', 'lodge', 'victorian']])
results = lm.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.566
Model:                            OLS   Adj. R-squared:                  0.566
Method:                 Least Squares   F-statistic:                     1963.
Date:                Sun, 31 Jan 2021   Prob (F-statistic):               0.00
Time:                        16:09:38   Log-Likelihood:                -85418.
No. Observations:                6028   AIC:                         1.708e+05
Df Residuals:                    6023   BIC:                         1.709e+05
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
intercept  -1.702e+05   1.68e+04    -10.148      0.000   -2.03e+05   -1.37e+05
bedrooms    1.654e+05   1.16e+04     14.295      0.000    1.43e+05    1.88e+05
bathrooms   1.125e+05   1.64e+04      6.844      0.000    8.03e+04    1.45e+05
lodge       1.637e+05   1.49e+04     10.965      0.000    1.34e+05    1.93e+05
victorian   6.713e+04   1.26e+04      5.342      0.000    4.25e+04    9.18e+04
==============================================================================
Omnibus:                      851.468   Durbin-Watson:                   2.007
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1342.460
Skew:                           0.984   Prob(JB):                    3.08e-292
Kurtosis:                       4.214   Cond. No.                         25.0
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [11]:
lm = sm.OLS(df2_new['price'], df2_new[['intercept', 'B', 'C', 'bedrooms', 'bathrooms', 'lodge', 'victorian']])
results = lm.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.809
Model:                            OLS   Adj. R-squared:                  0.809
Method:                 Least Squares   F-statistic:                     4250.
Date:                Sun, 31 Jan 2021   Prob (F-statistic):               0.00
Time:                        16:10:40   Log-Likelihood:                -82944.
No. Observations:                6028   AIC:                         1.659e+05
Df Residuals:                    6021   BIC:                         1.659e+05
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
intercept  -3.833e+05    1.2e+04    -31.995      0.000   -4.07e+05    -3.6e+05
B           5.229e+05   7040.928     74.271      0.000    5.09e+05    5.37e+05
C          -7168.6285   7639.254     -0.938      0.348   -2.21e+04    7807.045
bedrooms    1.732e+05   7677.152     22.558      0.000    1.58e+05    1.88e+05
bathrooms   9.996e+04   1.09e+04      9.164      0.000    7.86e+04    1.21e+05
lodge       1.685e+05   9906.629     17.012      0.000    1.49e+05    1.88e+05
victorian   7.056e+04   8337.790      8.463      0.000    5.42e+04    8.69e+04
==============================================================================
Omnibus:                      978.611   Durbin-Watson:                   1.993
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2926.472
Skew:                           0.848   Prob(JB):                         0.00
Kurtosis:                       5.962   Cond. No.                         25.9
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

80.9% of the variability in price can be explained by the linear model built using bedrooms, bathrooms, neighborhood, and home style.

For every additional bedroom a home has we can expect the price to increase by 173200, holding all other variables constant.

For every additional bathroom a home has we can expect the price to increase by 9960, holding all other variables constant.

We expect that a victorian house will cost 70560 more than a ranch house, all else being equal.

We expect that a house in neighborhood C will cost 7168 less than a neighborhood A house, all else being equal.